### PR TITLE
Feature/pf 152

### PR DESCRIPTION
**Pull request by Molvilada**
https://github.com/user-attachments/assets/0049ea6e-2be4-4985-9031-a04f799704de

## Release Notes

- **New Features**
  - Introduced a Room database for managing company data.
  - Added Kotlin Symbol Processing (KSP) support.
  - Implemented new UI components for error handling and snackbar customization.
  - Added a new `CompanyRepository` for better data management.

- **Improvements**
  - Enhanced localization with new error messages in Spanish and Portuguese.
  - Streamlined session state management in the main activity.
  - Updated dependency management for Room and KSP.

- **Bug Fixes**
  - Updated naming conventions for login request and response models for clarity.

- **Tests**
  - Added unit tests for the new `MainActivityViewModel` and updated tests for `LoginViewModel` and `SignUpViewModel`.

### CODE REVIEW

File: app/src/main/java/io/capibaras/abcall/ui/components/CustomSnackbarHost.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package io.capibaras.abcall.ui.components
+
+import androidx.compose.material3.Snackbar
+import androidx.compose.material3.SnackbarHost
+import androidx.compose.material3.SnackbarHostState
+import androidx.compose.runtime.Composable
+
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/ui/components/CustomSnackbarHost.kt:7]: Si son imports que no se usan el formateador los borra automaticamente

File: app/src/main/java/io/capibaras/abcall/viewmodels/SignUpViewModel.kt [hunk 2]
```diff
@@ -96,4 +105,24 @@ class SignUpViewModel : ViewModel() {
...
+        viewModelScope.launch {
+            isLoading = true
+            try {
+                val companyList = companyRepository.getCompanies(forceUpdate)
+                companies.value = companyList.map { it.name }
+            } catch (e: Exception) {
+                errorUIState = ErrorUIState.Error(
+                    if (e is IOException) R.string.error_network else R.string.error_get_companies
+                )
+            } finally {
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/viewmodels/SignUpViewModel.kt]: Ncesito poder cachear varios errores

File: app/src/main/java/io/capibaras/abcall/ui/views/SignUpScreen.kt [hunk 3]
```diff
@@ -52,127 +54,139 @@ import org.koin.androidx.compose.koinViewModel
...
-            verticalArrangement = Arrangement.Top
-        ) {
-            Image(
-                painter = painterResource(id = R.drawable.logo),
-                contentDescription = "Logo de ABCall",
-                modifier = Modifier
-                    .width(184.dp),
-                contentScale = ContentScale.Crop
-            )
+    val companies = viewModel.companies
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/ui/views/SignUpScreen.kt:68]: Si trato de dejarlo asi me dice que collectAsState Unresolved reference: collectAsState, y no me deja importarlo

File: app/src/main/java/io/capibaras/abcall/ui/MainActivity.kt [hunk 4]
```diff
@@ -4,44 +4,38 @@ import android.os.Bundle
...
-        } finally {
-            isSessionChecked = true
-        }
+@Composable
+fun MainScreen(viewModel: MainActivityViewModel = koinViewModel()) {
+    val isSessionChecked = viewModel.isSessionChecked
+    val isUserLoggedIn = viewModel.isUserLoggedIn
+
+    if (isSessionChecked) {
+        Navigation(isUserLoggedIn)
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/ui/MainActivity.kt:39]: En el viewmodel siempre devolverá algo

File: app/src/main/java/io/capibaras/abcall/data/repositories/CompanyRepository.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        val cacheExpired = System.currentTimeMillis() - lastUpdate > CACHE_EXPIRATION_TIME
+        val localData = companyDAO.getAllCompanies()
+
+        if (localData.isNotEmpty() && !forceUpdate && !cacheExpired) {
+            return localData
+        } else {
+            val remoteData = companyService.getCompanies()
+
+            companyDAO.deleteAllCompanies()
+            companyDAO.insertCompanies(remoteData)
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/data/repositories/CompanyRepository.kt]: Unresolved reference: runInTransaction, y no me deja importarlo

File: app/src/main/java/io/capibaras/abcall/viewmodels/LoginViewModel.kt [hunk 6]
```diff
@@ -69,36 +70,28 @@ class LoginViewModel(
...
-                                ErrorUIState.Error(R.string.error_authenticate)
-                            }
-                        } else {
-                            ErrorUIState.Error(R.string.error_authenticate)
-                        }
-                        ErrorUIState.Error(R.string.error_authenticate)
+                        val errorBody = response.errorBody()!!.string()
+                        val jsonObject = JSONObject(errorBody)
+                        val message = jsonObject.getString("message")
+                        ErrorUIState.Error(message = message)
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/viewmodels/LoginViewModel.kt:88]: errorbody no deberia ser null

File: app/src/main/java/io/capibaras/abcall/viewmodels/SignUpViewModel.kt [hunk 7]
```diff
@@ -5,19 +5,24 @@ import androidx.compose.runtime.getValue
...
-class SignUpViewModel : ViewModel() {
+class SignUpViewModel(private val companyRepository: CompanyRepository) : ViewModel() {
     var name by mutableStateOf("")
     var email by mutableStateOf("")
     var password by mutableStateOf("")
     var confirmPassword by mutableStateOf("")
-    var selectedText by mutableStateOf("")
-
-    var isLoading: Boolean = false
+    var company by mutableStateOf("")
```
[the remark this answers is not in the document]
Comment (Molvilada) [reply app/src/main/java/io/capibaras/abcall/viewmodels/SignUpViewModel.kt:21]: se puede modificar en el ui

**Review by VivianaOj — APPROVED** [review]

**Pull request merged**